FIAT/reference_element.py

```diff
@@ -1025,6 +1025,18 @@ def tuple_sum(tree):
         return tree
 
 
+def flatten_reference_cube(ref_el):
+    """Returns a flattened cube corresponding to a given tensor product cell."""
+    if numpy.sum(ref_el.get_dimension()) <= 1:
+        return ref_el
+    elif numpy.sum(ref_el.get_dimension()) == 2 and len(ref_el.vertices) == 4:
+        return UFCQuadrilateral()
+    elif numpy.sum(ref_el.get_dimension()) == 3 and len(ref_el.vertices) == 8:
+        return UFCHexahedron()
+    else:
+        raise TypeError('Invalid cell type')
+
```

Comment:
I think this is fine. Just for clarity, I believe the suggestion David and I made in our meeting would look something like this:
```python
def is_hypercube(cell):
    if isinstance(cell, (DefaultLine, UFCInterval, UFLQuadrilateral, UFLHexahedron)):
        return True
    elif isinstance(cell, TensorProductCell):
        for c in cell.cells:
            is_hypercube(c)
    else:
        return False

def flatten_reference_cube(ref_el):
    """ ... """
    flattened_cube = {2: UFCQuadrilateral(), 3: UFLHexahedron()}
    if numpy.sum(ref_el.get_dimension()) <= 1:
        # Just return point/interval cell arguments
        return ref_el
    else:
        # Handle cases where cell is a quad/cube constructed from a tensor product or
        # an already flattened element
        if is_hypercube(ref_el):
            return flattened_cube[numpy.sum(ref_el.get_dimension())]
        else:
            raise TypeError('Invalid cell type')
```
Note that I haven't ran this in code (so it might not be completely bullet-proof), but I think the idea is there.

Reply:
line 5 above should be: `return reduce(lambda a, b: a and b, [is_hypercube(c) for c in cell.cells])`

Reply:
AKA `reduce(operator.and_, map(is_hypercube, cell.cells))`

Reply:
(need to import reduce from functools)

Reply:
No. `operator.and_` is the bitwise operator. There is no function in the operator module corresponding to `and`.

Reply:
Oh yeah, sorry, always forget that idiocy.

Reply:
To be fair, on booleans there is no real difference between logical and bitwise and/or.